src/rdf4cpp/rdf/Literal.cpp

```diff
@@ -153,5 +153,341 @@ Literal Literal::make(std::string_view lexical_form, const IRI &datatype, Node::
+        std::any &rhs_value) {
+
+    using DatatypeEntry = datatypes::registry::DatatypeRegistry::DatatypeEntry;
+    using ConversionResult = datatypes::registry::DatatypeRegistry::ConversionResult;
+
+    auto const substitute_impl = [](auto const &to_substitute_entry, auto const &to_substitute_value, auto const &target_entry) {
+        DatatypeEntry result_entry = to_substitute_entry;
+        ConversionResult pr{ to_substitute_entry.datatype_iri, to_substitute_value };
+
+        while (result_entry.subtype_rank > target_entry.subtype_rank) {
+            pr = result_entry.into_supertype_fptr(pr.converted_value);
+            result_entry = datatypes::registry::DatatypeRegistry::get_entry(pr.converted_iri).value();
+        }
+
+        if (result_entry.subtype_rank != target_entry.subtype_rank) {
+            throw std::runtime_error{ "subtype substitution error, invalid hierarchy" };
+        }
+
+        return std::make_pair(result_entry, pr.converted_value);
+    };
+
+    if (lhs_entry.subtype_rank == 0 && rhs_entry.subtype_rank == 0) {
+        // nothing to do
+        return;
+    } else if (rhs_entry.subtype_rank > lhs_entry.subtype_rank) {
+        // substitute rhs
+        auto const [substituted_entry, substituted_value] = substitute_impl(rhs_entry, rhs_value, lhs_entry);
+
+        rhs_entry = substituted_entry;
+        rhs_value = substituted_value;
+    } else if (lhs_entry.subtype_rank > rhs_entry.subtype_rank) {
+        // substitute lhs
+        auto const [substituted_entry, substituted_value] = substitute_impl(lhs_entry, lhs_value, rhs_entry);
+
+        lhs_entry = substituted_entry;
+        lhs_value = substituted_value;
+    } else {
+        // substitute both once
+
+        if (lhs_entry.into_supertype_fptr == nullptr || rhs_entry.into_supertype_fptr == nullptr) {
+            throw std::runtime_error{ "subtype substitution unsupported for type" };
+        }
+
+        ConversionResult cr_lhs = lhs_entry.into_supertype_fptr(lhs_value);
+        ConversionResult cr_rhs = rhs_entry.into_supertype_fptr(rhs_value);
+
+        DatatypeEntry e_lhs = datatypes::registry::DatatypeRegistry::get_entry(cr_lhs.converted_iri).value();
+        DatatypeEntry e_rhs = datatypes::registry::DatatypeRegistry::get_entry(cr_rhs.converted_iri).value();
+
+        lhs_entry = e_lhs;
+        lhs_value = cr_lhs.converted_value;
+        rhs_entry = e_rhs;
+        rhs_value = cr_rhs.converted_value;
+    }
+}
+
+/**
+ * @brief use type promotion to try to convert lhs and rhs to the same type
+ *
+ * @param lhs_entry the datatype entry for the type of lhs
+ * @param lhs_value the value of lhs
+ * @param rhs_entry the datatype entry for the type of rhs
+ * @param rhs_value the value of rhs
+ */
+void promote(
+        datatypes::registry::DatatypeRegistry::DatatypeEntry &lhs_entry,
+        std::any &lhs_value,
+        datatypes::registry::DatatypeRegistry::DatatypeEntry &rhs_entry,
+        std::any &rhs_value) {
+
```

Comment:
```suggestion
assert(lhs_entry != rhs_entry);
```

Please also add this detail to the docstring